crates/linker/src/pe.rs

```diff
@@ -1159,83 +1174,135 @@ fn write_image_base_relocation(
+
+        let size_of_block = HEADER_WIDTH + relocations.len() * ENTRY_WIDTH;
+
+        header.virtual_address.set(LE, new_block_va);
+        header.size_of_block.set(LE, size_of_block as u32);
+
+        let number_of_entries = relocations.len();
+        let entries = load_structs_inplace_mut::<u16>(
+            mmap,
+            next_block_start as usize + HEADER_WIDTH,
+            number_of_entries,
+        );
+
+        entries[number_of_entries - relocations.len()..].copy_from_slice(relocations);
+
+        // sort by VA. Upper 4 bits store the relocation type
+        entries.sort_unstable_by_key(|x| x & 0b0000_1111_1111_1111);
+
+        size_of_block
+    }
 }
 
 /// the roc app functions are called from the host with an indirect call: the code will look
 /// in a table to find the actual address of the app function. This table must be relocated,
 /// because it contains absolute addresses to jump to.
 fn relocate_dummy_dll_entries(executable: &mut [u8], md: &PeMetadata) {
+    let thunks_start_va = (md.rdata_virtual_address - md.image_base as u32)
+        + md.thunks_start_offset_in_section as u32;
+
+    // relocations are defined per 4kb page
+    const BLOCK_SIZE: u32 = 4096;
+
+    let thunks_offset_in_block = thunks_start_va % BLOCK_SIZE;
+    let thunks_relocation_block_va = thunks_start_va - thunks_offset_in_block;
+
     let relocations: Vec<_> = (0..md.dynamic_relocations.name_by_virtual_address.len())
-        .map(|i| (md.thunks_start_offset_in_section + 2 * i) as u16)
+        .map(|i| (thunks_offset_in_block as usize + 2 * i) as u16)
         .collect();
 
     let added_reloc_bytes = write_image_base_relocation(
         executable,
         md.reloc_offset_in_file,
-        md.rdata_virtual_address - md.image_base as u32,
+        thunks_relocation_block_va,
         &relocations,
     );
 
     // the reloc section got bigger, and we need to update the header with the new size
     let reloc_section_header_start = md.dynamic_relocations.section_headers_offset_in_file as usize
         + md.reloc_section_index * std::mem::size_of::<ImageSectionHeader>();
 
-    let ptr = load_struct_inplace_mut::<ImageSectionHeader>(executable, reloc_section_header_start);
-    let new_virtual_size = ptr.virtual_size.get(LE) + added_reloc_bytes as u32;
-    ptr.virtual_size.set(LE, new_virtual_size);
+    let next_section = load_struct_inplace::<ImageSectionHeader>(
+        executable,
+        reloc_section_header_start + std::mem::size_of::<ImageSectionHeader>(),
+    );
+
+    let next_section_pointer_to_raw_data = next_section.pointer_to_raw_data.get(LE);
+
+    let reloc_section =
+        load_struct_inplace_mut::<ImageSectionHeader>(executable, reloc_section_header_start);
+    let old_section_size = reloc_section.virtual_size.get(LE);
+    let new_virtual_size = old_section_size + added_reloc_bytes as u32;
+    reloc_section.virtual_size.set(LE, new_virtual_size);
+
+    assert!(
```

Comment:
How likely is this to happen? Should it be essentially impossible?

Reply:
hmm no, it may happen if the reloc section just fills its current size. So we'll have to deal with that at some point. For now it just fails loudly